From da37a715f511e0a69d80111c390a44774f11ea39 Mon Sep 17 00:00:00 2001
From: czou <cao.zou@windriver.com>
Date: Thu, 3 Dec 2015 16:09:44 +0800
Subject: [PATCH 127/129] ls1021: add the support of RTLswitch for fixed-link
 property

RTLswitch phy not connected to the real MDIO bus, set the dts
fixed-link property of phy of RTLswitch, and create the fixed-link
phy in mach-ls1021.

Signed-off-by: czou <cao.zou@windriver.com>
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 arch/arm/boot/dts/ls1021a-iot.dts |    5 +----
 arch/arm/mach-imx/mach-ls1021a.c  |   31 +++++++++++++++++++++++++++++++
 2 files changed, 32 insertions(+), 4 deletions(-)

diff --git a/arch/arm/boot/dts/ls1021a-iot.dts b/arch/arm/boot/dts/ls1021a-iot.dts
index 23b28fb..675ae9b 100644
--- a/arch/arm/boot/dts/ls1021a-iot.dts
+++ b/arch/arm/boot/dts/ls1021a-iot.dts
@@ -108,7 +108,7 @@
 
 &enet2 {
 	tbi-handle = <&tbi1>;
-	phy-handle = <&phy2>;
+	fixed-link = <2 1 1000 0 0>;
 	phy-connection-type = "rgmii";
 	status = "ok";
 };
@@ -188,9 +188,6 @@
 		reg = <0x1>;
 		interrupts = <GIC_SPI 165 IRQ_TYPE_LEVEL_LOW>;
 	};
-	phy2: ethernet-phy@2 {
-		reg = <0x2>;
-	};
 	phy3: ethernet-phy@3 {
 		reg = <0x3>;
 		interrupts = <GIC_SPI 165 IRQ_TYPE_LEVEL_LOW>;
diff --git a/arch/arm/mach-imx/mach-ls1021a.c b/arch/arm/mach-imx/mach-ls1021a.c
index 7fc8076..42cf3c5 100644
--- a/arch/arm/mach-imx/mach-ls1021a.c
+++ b/arch/arm/mach-imx/mach-ls1021a.c
@@ -53,6 +53,37 @@ static void __init ls1021a_init_time(void)
 	tick_setup_hrtimer_broadcast();
 }
 
+#ifdef CONFIG_FIXED_PHY
+static int __init of_add_fixed_phys(void)
+{
+	int ret;
+	struct device_node *np;
+	u32 *fixed_link;
+	struct fixed_phy_status status = {};
+
+	for_each_node_by_name(np, "ethernet") {
+		fixed_link  = (u32 *)of_get_property(np, "fixed-link", NULL);
+		if (!fixed_link)
+			continue;
+
+		status.link = 1;
+		status.duplex = be32_to_cpu(fixed_link[1]);
+		status.speed = be32_to_cpu(fixed_link[2]);
+		status.pause = be32_to_cpu(fixed_link[3]);
+		status.asym_pause = be32_to_cpu(fixed_link[4]);
+
+		ret = fixed_phy_add(PHY_POLL, be32_to_cpu(fixed_link[0]), &status);
+		if (ret) {
+			of_node_put(np);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+arch_initcall(of_add_fixed_phys);
+#endif /* CONFIG_FIXED_PHY */
+
 static const char *ls1021a_dt_compat[] __initdata = {
 	"fsl,ls1021a",
 	NULL,
-- 
1.7.5.4

